### PR TITLE
${event-properties}: Added evaluateAsNestedProperties option

### PR DESCRIPTION
**Pull request by 304NotModified**
I was triggered by https://github.com/NLog/NLog/issues/3314,

example: 

    ${event-properties:key1.prop1:evaluateAsNestedProperties=true}

As (maybe naive) implementation to support this.

Choices to be made:

1. Is this a suitable feature and place? (${event-properties})
2. How should I name this? EvaluateAsNestedProperties is from NLog.Web
3. Is this config logical, or should we configure the path separately? e.g. `${event-properties:key1:path=prop1}`
4. Should be support arrays in the future? (e.g.`key1[prop1]`)
5. Do we need (fancy) caching (now or in the future)? Is the `ObjectReflectionCache` suitable for this?
6. What do you think about the case-sensitive-case-insensitive lookup? (best for configuration, not for performance, although we could improve that with 2 dictionaries)

@snakefoot you opinion please :)

### CODE REVIEW

**Comment by snakefoot** [comment]
I would use the ObjectReflectionCache. It makes it very easy to enumerate properties (And caching and provides compiled fast lambda optimization). Maybe improve the enumerator to perform single property lookup (or maybe just cache the property lambda lookup on the layoutrenderer)


I also considered using ${event properties} but thought it would become too wordy because one had to specify extra option. I like it more terse. Ex ${eventobjectproperty=myOrder.OrderId}.


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


I was triggered by #3314<https://github.com/NLog/NLog/issues/3314>,

example:

${event-properties:key1.prop1:evaluateAsNestedProperties=true}

As (maybe naive) implementation to support this.

Choices to be made:

  1.  Is this a suitable feature and place? (${event-properties})
  2.  How should I name this? EvaluateAsNestedProperties is from NLog.Web
  3.  Is this config logical, or should we splits the path? e.g. ${event-properties:key1:path=prop1}
  4.  Should be support arrays in the future? (e.g. key1[prop1]
  5.  Do we need fancy caching (now or in the future)? Is the ObjectReflectionCache suitable for this?

@snakefoot<https://github.com/snakefoot> you opinion please :)

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/NLog/NLog/pull/3318

Commit Summary

  *   ${event-properties}: Added evaluateAsNestedProperties option

File Changes

  *   A src/NLog/Internal/PropertyReader.cs<https://github.com/NLog/NLog/pull/3318/files#diff-0> (75)
  *   M src/NLog/LayoutRenderers/EventPropertiesLayoutRenderer.cs<https://github.com/NLog/NLog/pull/3318/files#diff-1> (62)
  *   M tests/NLog.UnitTests/LayoutRenderers/EventPropertiesTests.cs<https://github.com/NLog/NLog/pull/3318/files#diff-2> (19)

Patch Links:

  *   https://github.com/NLog/NLog/pull/3318.patch
  *   https://github.com/NLog/NLog/pull/3318.diff

**Comment by snakefoot** [comment]
Created #3329 as an example of fancy caching.

**Comment by snakefoot** [comment]
I guess one could consider adding extra stuff:
- Object Dictionary Lookup
- Taking First-Item in an Array/List/Enumerable

**Comment by 304NotModified** [comment]
superseded by #3329
